rx_signal_processing/decimate.cu

```diff
@@ -384,8 +434,260 @@ void decimate2048_wrapper(cuComplex* original_samples,
   auto shr_mem_taps = num_freqs * num_taps_per_filter * sizeof(cuComplex);
   std::cout << "    Number of shared memory bytes: "<< shr_mem_taps << std::endl;
 
-  auto dimGrid = create_grid(samples_per_antenna, dm_rate, num_antennas);
-  auto dimBlock = create_block(num_taps_per_filter/2, num_freqs);
-  decimate2048<<<dimGrid,dimBlock,shr_mem_taps,stream>>>(original_samples, decimated_samples,
+  auto dimGrid = create_bandpass_grid(samples_per_antenna, dm_rate, num_antennas);
+  auto dimBlock = create_bandpass_block(num_taps_per_filter/2, num_freqs);
+  bandpass_decimate2048<<<dimGrid,dimBlock,shr_mem_taps,stream>>>(original_samples, decimated_samples,
     filter_taps, dm_rate, samples_per_antenna);
+}
+
+/**
+ * @brief      Performs decimation using a lowpass filter on one or more sets of baseband samples 
+ * corresponding to each RX frequency. This algorithm works on filters with less that 1024 taps.
+ *
+ * @param[in]  original_samples     A pointer to input samples for one or more baseband datasets.
+ * @param[in]  decimated_samples    A pointer to a buffer to place output samples for each frequency
+ *                                  dataset after decimation.
+ * @param[in]  filter_taps          A pointer to a lowpass filter used for further decimation.
+ * @param[in]  dm_rate              Decimation rate.
+ * @param[in]  samples_per_antenna  The number of samples per antenna in the original set of
+ *                                  samples.
+ * 
+ * This function performs a parallel version of filtering+downsampling on the GPU to be able
+ * process data in realtime. This algorithm will use 1 GPU thread per filter tap if there are less
+ * than 1024 taps for all filters combined. Only works with power of two length filters, or a
+ * filter that is zero padded to a power of two in length. This algorithm takes one or more 
+ * baseband datasets corresponding to each RX frequency and filters each one using a single lowpass
+ * filter before downsampling.
+ * 
+ *   gridDim.x - The number of decimated output samples for one frequency set.
+ *   gridDim.y - Total number of antennas.
+ *   gridDim.z - Total number of frequency data sets.
+ *
+ *   blockIdx.x - Decimated output sample index.
+ *   blockIdx.y - Antenna index.
+ *   blockIdx.z - Frequency dataset index.
+ *
+ *   blockDim.x - Number of filter taps in each filter.
+
+ *   threadIdx.x - Filter tap indices.
+ */
+__global__ void lowpass_decimate1024(cuComplex* original_samples,
+  cuComplex* decimated_samples,
+  cuComplex* filter_taps, uint32_t dm_rate,
+  uint32_t samples_per_antenna) {
+
+  // Since number of filter taps is calculated at runtime and we do not want to hardcode
+  // values, the shared memory can be dynamically initialized at invocation of the kernel.
+  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared
+
+  extern __shared__ cuComplex filter_products[];
+
+  auto frequency_dataset_offset = blockIdx.z * (gridDim.x * dm_rate);
```

Comment:
We don't know if the memory layout is grouped by freq or antenna so can't say for sure if this calculation is correct.